### PR TITLE
[SDP-1076] httphandler/services/data: publish the receiver wallets invitation events on disbursement start

### PR DESCRIPTION
**Pull request by CaioTeixeira95**
### What

This PR moves the receiver wallets' publishing events from the `data.DisbursementInstructionModel.ProccessAll` function to `services.DisbursementManagerService.DisbursementStart`.

### Why

Since uploading instructions may happen multiple times before a disbursement starts, issuing that event there could cause unnecessary spamming to receivers.

### Known limitations

N/A

#### PR Structure

* [x] This PR has a reasonably narrow scope (if not, break it down into smaller PRs).
* [x] This PR title and description are clear enough for anyone to review it.
* [x] This PR does not mix refactoring changes with feature changes (split into two PRs otherwise).

#### Thoroughness

* [x] This PR adds tests for the new functionality or fixes.
* [x] This PR contains the link to the Jira ticket it addresses.

#### Configs and Secrets

* [x] No new CONFIG variables are required -OR- the new required ones were added to the helmchart's [`values.yaml`] file.
* [x] No new CONFIG variables are required -OR- the new required ones were added to the deployments ([`pr-preview`], [`dev`], [`demo`], `prd`).
* [x] No new SECRETS variables are required -OR- the new required ones were mentioned in the helmchart's [`values.yaml`] file.
* [x] No new SECRETS variables are required -OR- the new required ones were added to the deployments ([`pr-preview secrets`], [`dev secrets`], [`demo secrets`], `prd secrets`).

#### Release

* [x] This is not a breaking change.
* [ ] **This is ready for production.**. If your PR is not ready for production, please consider opening additional complementary PRs using this one as the base. Only merge this into `develop` or `main` after it's ready for production!

#### Deployment

* [ ] Does the deployment work after merging?

[`values.yaml`]: ../helmchart/sdp/values.yaml
[`pr-preview`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/common-previews/stellar-disbursement-platform/backend-helm-values
[`dev`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/backend-helm-values
[`demo`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/demo/demo-backend-helm-values
[`pr-preview secrets`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/common-previews/externalsecrets-common-previews.yaml#L241-L346
[`dev secrets`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/stellar-disbursement-platform-externalsecrets.yaml
[`demo secrets`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/demo/demo-sdp-externalsecrets.yaml

### CODE REVIEW

**Review by marwen-abid — APPROVED** [review]
LGTM! Good to merge after addressing the issue mentioned below!

File: internal/services/disbursement_management_service.go [hunk 1]
```diff
@@ -149,41 +149,79 @@ func (s *DisbursementManagementService) StartDisbursement(ctx context.Context, d
...
+		}
+
+		// If there's no message to publish we only log it.
+		if len(msgs) == 0 {
+			log.Ctx(ctx).Infof("no messages to be published for disbursement ID %s", disbursementID)
+			return nil
+		}
+
+		if err = s.produceEvents(ctx, msgs...); err != nil {
+			return err
```
Comment (marcelosalloum) [inline internal/services/disbursement_management_service.go:205]: Can you wrap this error before returning it, plz?

File: internal/services/disbursement_management_service.go [hunk 2]
```diff
@@ -149,41 +149,79 @@ func (s *DisbursementManagementService) StartDisbursement(ctx context.Context, d
...
 		}
 
 		return nil
 	})
 }
 
+func (s *DisbursementManagementService) produceEvents(ctx context.Context, msgs ...events.Message) error {
+	if s.eventProducer == nil {
+		log.Ctx(ctx).Errorf("event producer is nil, could not publish messages %s", msgs)
```
Comment (marcelosalloum) [inline internal/services/disbursement_management_service.go:214]: This small change should provide more info when debugging the logs:

```suggestion
		log.Ctx(ctx).Errorf("event producer is nil, could not publish messages %+v", msgs)
```

File: internal/data/receivers_wallet.go [hunk 3]
```diff
@@ -224,34 +224,54 @@ func (rw *ReceiverWalletModel) GetByReceiverIDsAndWalletID(ctx context.Context,
...
+`
+
+var getPendingRegistrationReceiverWalletsBaseArgs = []any{ReadyReceiversWalletStatus}
+
+func (rw *ReceiverWalletModel) GetAllPendingRegistrationByReceiverWalletIDs(ctx context.Context, sqlExec db.SQLExecuter, receiverWalletIDs []string) ([]*ReceiverWallet, error) {
+	query := fmt.Sprintf(getPendingRegistrationReceiverWalletsBaseQuery, "AND rw.id = ANY($2)")
 
 	receiverWallets := make([]*ReceiverWallet, 0)
-	err := rw.dbConnectionPool.SelectContext(ctx, &receiverWallets, query, pq.Array(receiverWalletIDs))
+	args := append(getPendingRegistrationReceiverWalletsBaseArgs, pq.Array(receiverWalletIDs))
```
[the remark this answers is not in the document]
Comment (CaioTeixeira95) [reply internal/data/receivers_wallet.go:257]: Thank you for pointing it out. I've made a test for this case: [b811f6d](https://github.com/stellar/stellar-disbursement-platform-backend/pull/182/commits/b811f6d32362864a15cb42b3efce11fdfa16b653)

File: internal/data/receivers_wallet.go [hunk 4]
```diff
@@ -224,34 +224,58 @@ func (rw *ReceiverWalletModel) GetByReceiverIDsAndWalletID(ctx context.Context,
...
+		INNER JOIN wallets w ON w.id = rw.wallet_id
+		INNER JOIN disbursements d ON w.id = d.wallet_id
+		INNER JOIN payments p ON d.id = p.disbursement_id AND p.receiver_id = r.id
+	WHERE
+		rw.status = $1 -- 'READY'::receiver_wallet_status
+		%s
+	GROUP BY
+		rw.id,
+		r.id,
+		w.id
```
Comment (marcelosalloum) [inline internal/data/receivers_wallet.go:248]: Isn't it enough to group by receiver_wallet.id?

Comment (CaioTeixeira95) [reply internal/data/receivers_wallet.go:248]: It isn't. Postgres requires all these fields on the group by.

**Pull request merged**